compiler/rustc_trait_selection/src/traits/error_reporting/mod.rs

```diff
@@ -1621,6 +1621,36 @@ impl<'a, 'tcx> InferCtxtPrivExt<'a, 'tcx> for InferCtxt<'a, 'tcx> {
                 "type mismatch resolving `{}`",
                 predicate
             );
+            let mut v = HirFinder { needle: obligation.cause.span, found: None };
+            let body = self.tcx.hir().get(obligation.cause.body_id);
+            if let hir::Node::Expr(expr) = body {
+                v.visit_expr(expr);
+            }
+            if let Some(expr) = v.found {
+                // FIXME(estebank): look for method calls that can be applied on the previous
+                // element of the chain, that would make the next call to typeck.
```

Comment:
@compiler-errors I think that this is the part that would be useful from this PR, that this PR didn't handle: enabling rustc to look for method calls that could be added to make the code compile would be the biggest benefit of this change, but this PR on its own might not be as useful as I envisioned.

Reply:
At least in the case of #33941, suggesting other methods that are compatible with the real `Item` type doesn't seem like the right direction for a suggestion to go imo. 

I _think_ the compiler should be trying to stress that that `(&K, &V)` is not `&T` instead. I'm not sure how to make the connection more clear that if they want `(&K, &V) -> (K, V)` they've gotta do the map themselves... now I don't think this needs special-casing per se, but conversely, I think suggesting other methods is probably going to lead them further astray?